File: src/management/stats.py
import asyncio
from datetime import datetime, timedelta

from telegram import Message, Update
from telegram.constants import ParseMode
from telegram.ext import ContextTypes
from telegram.helpers import mention_html

import commands
import utils.string
from config.db import get_db
from utils import readable_time
from utils.decorators import description, example, triggers, usage
from utils.messages import get_message

_DOW_NAMES = ("Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat")


async def stat_string_builder(
    rows: list,
    message: Message,
    context: ContextTypes.DEFAULT_TYPE,
    total_count: int,
) -> None:
    if not rows:
        await message.reply_text("No messages recorded.")
        return

    # Parallel fetch all names (cache hits are instant, misses run concurrently)
    user_ids = [user_id for _, user_id, _ in rows]
    names = await asyncio.gather(
        *(utils.string.get_first_name(uid, context) for uid in user_ids)
    )

    # Build output with list join (avoids O(n²) string concat)
    lines = [f"Stats for <b>{message.chat.title}:</b>", ""]
    for (_, _user_id, count), name in zip(rows, names, strict=True):
        percent = count / total_count * 100
        lines.append(f"<code>{percent:4.1f}% - {name}</code>")
    lines.append(f"\nTotal messages: <b>{total_count}</b>")

    await message.reply_text("\n".join(lines), parse_mode=ParseMode.HTML)


@triggers(["seen"])
@usage("/seen [username]")
@example("/seen @obviyus")
@description("Get duration since last message of a user.")
async def get_last_seen(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    message = get_message(update)
    if not message:
        return
    if not context.args:
        await commands.usage_string(message, get_last_seen)
        return

    username_input = context.args[0].split("@")
    if len(username_input) <= 1:
        await commands.usage_string(message, get_last_seen)
        return

    username_lower = username_input[1].lower()

    async with get_db() as conn:
        async with conn.execute(
            "SELECT username, last_seen, last_message_link FROM user_stats WHERE LOWER(username) = ?",
            (username_lower,),
        ) as cursor:
            user_stats = await cursor.fetchone()

    if not user_stats:
        await message.reply_text(f"@{username_input[1]} has never been seen.")
        return

    last_seen = user_stats["last_seen"]
    message_link = user_stats["last_message_link"]
    username_display = user_stats["username"]

    try:
        last_seen_int = int(last_seen)
    except ValueError:
        last_seen_int = int(datetime.fromisoformat(last_seen).timestamp())

    # Create a link to the message if available
    html_message = f"\n\n🔗 <a href='{message_link}'>Link</a>" if message_link else ""

    user_mention_string = mention_html(username_display, username_display)
    await message.reply_text(
        f"Last message from {user_mention_string} was {await readable_time(last_seen_int)} ago.{html_message}",
        disable_web_page_preview=True,
        disable_notification=True,
        parse_mode=ParseMode.HTML,
    )


def _sparkline(values: list[int]) -> str:
    """Return a unicode sparkline for a list of non-negative integers."""
    if not values:
        return ""
    levels = "▁▂▃▄▅▆▇█"
    max_value = max(values) or 1
    return "".join(
        levels[min(len(levels) - 1, int(v / max_value * (len(levels) - 1)))]
        for v in values
    )


def _format_horizontal_bars(
    labels: list[str], values: list[int], width: int = 20
) -> str:
    """Return a simple monospaced horizontal bar chart string."""
    if not values:
        return ""
    max_value = max(values) or 1
    lines: list[str] = []
    for label, value in zip(labels, values, strict=False):
        bar_len = 0 if max_value == 0 else round(value / max_value * width)
        bar = "█" * bar_len
        lines.append(f"{label:>3} | {bar:<{width}} {value}")
    return "\n".join(lines)


async def _resolve_target_user(
    update: Update, context: ContextTypes.DEFAULT_TYPE
) -> tuple[int | None, str | None]:
    """Resolve target user_id and display username from args/reply/self."""
    message = get_message(update)

    # Priority 1: explicit @username arg → DB lookup
    if context.args:
        username_input = context.args[0].split("@")
        if len(username_input) > 1:
            username_lower = username_input[1].lower()
            async with get_db() as conn:
                async with conn.execute(
                    "SELECT user_id, username FROM user_stats WHERE LOWER(username) = ?",
                    (username_lower,),
                ) as cursor:
                    row = await cursor.fetchone()
            if row:
                return int(row["user_id"]), row["username"]

    # Priority 2: reply target, else caller (unified extraction)
    user = (
        message.reply_to_message.from_user
        if message and message.reply_to_message
        else None
    ) or (message.from_user if message else None)
    if user:
        return user.id, user.username or user.full_name

    return None, None


async def _fetch_chat_stats(chat_id: int, today_only: bool = False) -> tuple[list, int]:
    """Fetch top users and total count, optionally filtered to today."""
    time_clause = (
        "AND create_time >= DATE('now', 'localtime') AND create_time < DATE('now', '+1 day', 'localtime')"
        if today_only
        else ""
    )

    async with get_db() as conn:
        async with conn.execute(
            f"""
            SELECT create_time, user_id, COUNT(user_id) AS user_count
            FROM chat_stats
            WHERE chat_id = ? {time_clause}
            GROUP BY user_id
            ORDER BY COUNT(user_id) DESC
            LIMIT 10;
            """,
            (chat_id,),
        ) as cursor:
            users = list(await cursor.fetchall())

        async with conn.execute(
            f"""
            SELECT COUNT(id) AS total_count
            FROM chat_stats
            WHERE chat_id = ? {time_clause};
            """,
            (chat_id,),
        ) as cursor:
            result = await cursor.fetchone()
            total_count: int = result["total_count"] if result else 0

    return users, total_count


@usage("/stats")
@example("/stats")
@triggers(["stats"])
@description("Get message count by user for the last day.")
async def get_chat_stats(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    message = get_message(update)
    if not message:
        return
    users, total = await _fetch_chat_stats(message.chat_id, today_only=True)
    await stat_string_builder(users, message, context, total)


@usage("/gstats")
@example("/gstats")
@triggers(["gstats"])
@description("Get total message count by user of this group.")
async def get_total_chat_stats(
    update: Update, context: ContextTypes.DEFAULT_TYPE
) -> None:
    message = get_message(update)
    if not message:
        return
    users, total = await _fetch_chat_stats(message.chat_id, today_only=False)
    await stat_string_builder(users, message, context, total)


@usage("/ustats [username]")
@example("/ustats @obviyus")
@triggers(["ustats", "ustat", "userstats"])
@description("Show a user's weekly activity and other stats in this group.")
async def get_user_stats(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    message = get_message(update)
    if not message:
        return
    chat_id = message.chat_id

    user_id, username_display = await _resolve_target_user(update, context)
    if not user_id:
        await message.reply_text(
            "Couldn't resolve target user. Usage: /ustats [@username]"
        )
        return

    # Build the last 7 days (oldest -> newest) date keys in localtime
    today_local = datetime.now()
    day_keys: list[str] = []
    day_labels: list[str] = []
    for i in range(6, -1, -1):
        day = today_local - timedelta(days=i)
        day_keys.append(day.strftime("%Y-%m-%d"))
        day_labels.append(day.strftime("%a"))

    # AIDEV-NOTE: Single CTE query replaces 6 sequential DB round-trips
    async with get_db() as conn:
        async with conn.execute(
            """
            WITH user_daily AS (
                SELECT DATE(create_time, 'localtime') AS d, COUNT(*) AS cnt
                FROM chat_stats
                WHERE chat_id = ? AND user_id = ?
                  AND create_time >= DATE('now', '-6 days', 'localtime')
                  AND create_time < DATE('now', '+1 day', 'localtime')
                GROUP BY d
            ),
            group_week AS (
                SELECT COUNT(*) AS total
                FROM chat_stats
                WHERE chat_id = ?
                  AND create_time >= DATE('now', '-6 days', 'localtime')
                  AND create_time < DATE('now', '+1 day', 'localtime')
            ),
            top_hour AS (
                SELECT strftime('%H', create_time, 'localtime') AS hour
                FROM chat_stats
                WHERE chat_id = ? AND user_id = ?
                  AND create_time >= DATE('now', '-30 days', 'localtime')
                GROUP BY hour
                ORDER BY COUNT(*) DESC
                LIMIT 1
            ),
            top_dow AS (
                SELECT strftime('%w', create_time, 'localtime') AS dow
                FROM chat_stats
                WHERE chat_id = ? AND user_id = ?
                  AND create_time >= DATE('now', '-90 days', 'localtime')
                GROUP BY dow
                ORDER BY COUNT(*) DESC
                LIMIT 1
            ),
            lifetime AS (
                SELECT COUNT(*) AS total
                FROM chat_stats
                WHERE chat_id = ? AND user_id = ?
            )
            SELECT
                'daily' AS qtype, d AS val, cnt AS cnt FROM user_daily
            UNION ALL
            SELECT 'group_total', NULL, total FROM group_week
            UNION ALL
            SELECT 'top_hour', hour, NULL FROM top_hour
            UNION ALL
            SELECT 'top_dow', dow, NULL FROM top_dow
            UNION ALL
            SELECT 'lifetime', NULL, total FROM lifetime
            """,
            (
                chat_id,
                user_id,
                chat_id,
                chat_id,
                user_id,
                chat_id,
                user_id,
                chat_id,
                user_id,
            ),
        ) as cursor:
            rows = await cursor.fetchall()

        # Parse unified result set
        counts_map: dict[str, int] = {}
        group_week_total = 0
        top_hour = None
        top_dow = None
        lifetime_total = 0

        for row in rows:
            qtype = row["qtype"]
            if qtype == "daily":
                counts_map[row["val"]] = row["cnt"]
            elif qtype == "group_total":
                group_week_total = row["cnt"] or 0
            elif qtype == "top_hour":
                top_hour = row["val"]
            elif qtype == "top_dow":
                top_dow = row["val"]
            elif qtype == "lifetime":
                lifetime_total = row["cnt"] or 0

        week_counts = [counts_map.get(k, 0) for k in day_keys]
        user_week_total = sum(week_counts)

        # Streak calculation needs active dates (separate query - still faster than 6)
        async with conn.execute(
            """
            SELECT DATE(create_time, 'localtime') AS d
            FROM chat_stats
            WHERE chat_id = ? AND user_id = ?
              AND create_time >= DATE('now', '-60 days', 'localtime')
            GROUP BY d
            """,
            (chat_id, user_id),
        ) as cursor:
            date_rows = await cursor.fetchall()
            active_dates = {r["d"] for r in date_rows}

    # Streak calculation from today backwards
    streak = 0
    probe = today_local
    while streak < 60:  # cap
        key = probe.strftime("%Y-%m-%d")
        if key in active_dates:
            streak += 1
            probe -= timedelta(days=1)
        else:
            break

    # Build charts
    spark = _sparkline(week_counts)
    bars = _format_horizontal_bars(day_labels, week_counts)

    # Friendly labels
    top_dow_name = _DOW_NAMES[int(top_dow)] if top_dow is not None else "-"
    top_hour_label = (
        f"{int(top_hour):02d}:00-{(int(top_hour) + 1) % 24:02d}:00"
        if top_hour is not None
        else "-"
    )

    share = (
        (user_week_total / group_week_total * 100.0) if group_week_total > 0 else 0.0
    )
    avg_per_day = user_week_total / 7.0

    # Header name
    target_name = username_display or str(user_id)
    header = f"User stats for <b>{target_name}</b> in <b>{message.chat.title}</b>\n"

    if user_week_total == 0:
        await message.reply_text(
            header + "\nNo messages in the last 7 days.",
            parse_mode=ParseMode.HTML,
        )
        return

    text = (
        header
        + f"\nLast 7d: <b>{user_week_total}</b> msgs ({share:.1f}% of group) • avg {avg_per_day:.2f}/day"
        + f"\nLifetime in this chat: <b>{lifetime_total}</b> msgs"
        + (
            f"\nMost active hour: <b>{top_hour_label}</b> \nMost active day: <b>{top_dow_name}</b>"
        )
        + (f"\nCurrent streak: <b>{streak}</b> day(s)" if streak > 0 else "")
        + "\n\n<pre>"
        + bars
        + "</pre>"
        + (f"\nSparkline: <code>{spark}</code>" if spark else "")
    )

    await message.reply_text(text, parse_mode=ParseMode.HTML)

File: src/features/stats.ts
import { Effect, getMe, html } from "telly";

import {
  answer,
  type CommandDefinition,
  usage,
} from "../app/command.ts";
import { rowNumber, rowString } from "../app/database.ts";
import type { AppDependencies } from "../app/dependencies.ts";

function readableTime(now: Date, input: string): string {
  const seconds = Math.abs(now.getTime() - new Date(input).getTime()) / 1_000;
  for (const [limit, unit, divisor] of [
    [60, "second", 1],
    [3_600, "minute", 60],
    [86_400, "hour", 3_600],
    [604_800, "day", 86_400],
    [31_536_000, "week", 604_800],
    [Number.POSITIVE_INFINITY, "year", 31_536_000],
  ] as const) {
    if (seconds >= limit) continue;
    const value = Number((seconds / divisor).toFixed(1));
    return `${value} ${unit}${value > 1 ? "s" : ""}`;
  }
  return "0 seconds";
}

function statsCommand(dependencies: AppDependencies, todayOnly: boolean): CommandDefinition {
  return {
    description: todayOnly
      ? "Get message count by user for the last day."
      : "Get total message count by user in this group.",
    example: todayOnly ? "/stats" : "/gstats",
    names: [todayOnly ? "stats" : "gstats"],
    run: Effect.fn(todayOnly ? "dailyStats" : "groupStats")(function* (match) {
      const time = todayOnly
        ? "AND create_time >= DATE('now', 'localtime') AND create_time < DATE('now', '+1 day', 'localtime')"
        : "";
      const rows = yield* dependencies.database.all(
        `SELECT cs.user_id, COUNT(*) AS user_count,
                COALESCE(us.first_name, us.username, CAST(cs.user_id AS TEXT)) AS name
         FROM chat_stats cs
         LEFT JOIN user_stats us ON us.user_id = cs.user_id
         WHERE cs.chat_id = ? ${time}
         GROUP BY cs.user_id
         ORDER BY COUNT(*) DESC
         LIMIT 10`,
        [match.message.chat.id],
      );
      const total = rows.reduce((sum, row) => sum + rowNumber(row, "user_count"), 0);
      if (total === 0) return yield* answer(match.message, "No messages recorded.");
      const title = match.message.chat.title ?? String(match.message.chat.id);
      const lines = rows.map((row) => {
        const count = rowNumber(row, "user_count");
        return `<code>${(count / total * 100).toFixed(1).padStart(4)}% - ${html.escape(rowString(row, "name"))}</code>`;
      });
      return yield* answer(match.message, {
        parseMode: "HTML",
        text: `Stats for <b>${html.escape(title)}:</b>\n\n${lines.join("\n")}\n\nTotal messages: <b>${total}</b>`,
      });
    }),
    usage: todayOnly ? "/stats" : "/gstats",
  };
}

function seenCommand(dependencies: AppDependencies): CommandDefinition {
  const definition: CommandDefinition = {
    description: "Get the time since a user's last message in this chat.",
    example: "/seen @obviyus",
    names: ["seen"],
    run: Effect.fn("seen")(function* (match) {
      const username = match.args[0]?.replace(/^@/u, "").toLowerCase();
      if (username === undefined || username.length === 0) {
        return yield* usage(match.message, definition);
      }
      const row = yield* dependencies.database.one(
        `SELECT us.user_id, us.username, cs.message_id, cs.create_time
         FROM user_stats us
         JOIN chat_stats cs ON cs.user_id = us.user_id AND cs.chat_id = ?
         WHERE LOWER(us.username) = ?
         ORDER BY cs.id DESC LIMIT 1`,
        [match.message.chat.id, username],
      );
      if (row === undefined) {
        return yield* answer(match.message, `@${username} has never been seen in this chat.`);
      }
      const messageId = rowNumber(row, "message_id");
      const chatId = String(match.message.chat.id);
      const messageLink = match.message.chat.username === undefined && chatId.startsWith("-100")
        ? `https://t.me/c/${chatId.slice(4)}/${messageId}`
        : match.message.chat.username === undefined
        ? undefined
        : `https://t.me/${match.message.chat.username}/${messageId}`;
      const link = messageLink === undefined ? "" : `\n\n🔗 <a href="${messageLink}">Link</a>`;
      return yield* answer(match.message, {
        linkPreviewOptions: { isDisabled: true },
        parseMode: "HTML",
        text: `Last message from <a href="tg://user?id=${rowNumber(row, "user_id")}">${html.escape(rowString(row, "username"))}</a> was ${readableTime(dependencies.now(), rowString(row, "create_time"))} ago.${link}`,
      });
    }),
    usage: "/seen [username]",
  };
  return definition;
}

function totalCommand(dependencies: AppDependencies, users: boolean): CommandDefinition {
  return {
    description: users
      ? "Get the number of users that use this bot."
      : "Get the number of groups that use this bot.",
    example: users ? "/users" : "/groups",
    names: [users ? "users" : "groups"],
    run: Effect.fn(users ? "totalUsers" : "totalGroups")(function* (match) {
      const bot = yield* getMe();
      const row = yield* dependencies.database.one(
        `SELECT COUNT(DISTINCT ${users ? "user_id" : "chat_id"}) AS total FROM chat_stats`,
      );
      const total = row === undefined ? 0 : rowNumber(row, "total");
      return yield* answer(match.message, {
        parseMode: "HTML",
        text: users
          ? `@${bot.username ?? bot.firstName} is used by <b>${total}</b> users.`
          : `@${bot.username ?? bot.firstName} is used in <b>${total}</b> groups.`,
      });
    }),
    usage: users ? "/users" : "/groups",
  };
}

function botStatsCommand(dependencies: AppDependencies): CommandDefinition {
  return {
    description: "Get usage counts for bot commands.",
    example: "/botstats",
    names: ["botstats"],
    run: Effect.fn("botStats")(function* (match) {
      const bot = yield* getMe();
      const rows = yield* dependencies.database.all(
        `SELECT command, COUNT(*) AS command_count
         FROM command_stats GROUP BY command ORDER BY command_count DESC LIMIT 10`,
      );
      const totalRow = yield* dependencies.database.one("SELECT COUNT(*) AS total FROM command_stats");
      const lines = rows.map((row) =>
        `<code>${String(rowNumber(row, "command_count")).padStart(4)} - /${html.escape(rowString(row, "command"))}</code>`
      );
      return yield* answer(match.message, {
        parseMode: "HTML",
        text: `Stats for <b>@${html.escape(bot.username ?? bot.firstName)}:</b>\n\n${lines.join("\n")}\n\nTotal: <b>${totalRow === undefined ? 0 : rowNumber(totalRow, "total")}</b>`,
      });
    }),
    usage: "/botstats",
  };
}

function friendsCommand(dependencies: AppDependencies): CommandDefinition {
  return {
    description: "Show your strongest connections in this group's social graph.",
    example: "/friends",
    names: ["friends"],
    run: Effect.fn("friends")(function* (match) {
      const user = match.message.from;
      if (user === undefined) return;
      const rows = yield* dependencies.database.all(
        `SELECT other_id, direction, SUM(weight) AS weight,
                COALESCE(us.first_name, us.username, CAST(other_id AS TEXT)) AS name
         FROM (
           SELECT mentioned_user_id AS other_id, 'out' AS direction, COUNT(*) AS weight
           FROM chat_mentions
           WHERE chat_id = ? AND mentioning_user_id = ? AND mentioned_user_id != ?
           GROUP BY mentioned_user_id
           UNION ALL
           SELECT mentioning_user_id AS other_id, 'in' AS direction, COUNT(*) AS weight
           FROM chat_mentions
           WHERE chat_id = ? AND mentioned_user_id = ? AND mentioning_user_id != ?
           GROUP BY mentioning_user_id
         ) edges
         LEFT JOIN user_stats us ON us.user_id = other_id
         GROUP BY other_id, direction
         ORDER BY weight DESC`,
        [
          match.message.chat.id,
          user.id,
          user.id,
          match.message.chat.id,
          user.id,
          user.id,
        ],
      );
      if (rows.length === 0) {
        const graph = yield* dependencies.database.one(
          "SELECT 1 FROM chat_mentions WHERE chat_id = ? LIMIT 1",
          [match.message.chat.id],
        );
        return yield* answer(
          match.message,
          graph === undefined
            ? "This group has no social graph yet."
            : "You are not in this group's social graph.",
        );
      }
      const outgoing = rows.filter((row) => row["direction"] === "out").slice(0, 3);
      const incoming = rows.filter((row) => row["direction"] === "in").slice(0, 3);
      const render = (heading: string, arrow: string, values: typeof rows) => values.length === 0
        ? ""
        : `\n\n${heading}${values.map((row) => `\n<code>${String(rowNumber(row, "weight")).padStart(6)} ${arrow} ${html.escape(rowString(row, "name"))}</code>`).join("")}`;
      return yield* answer(match.message, {
        parseMode: "HTML",
        text: `From the social graph of <b>${html.escape(match.message.chat.title ?? String(match.message.chat.id))}</b>:${render("You have the strongest connections to:", "⟶", outgoing)}${render("You have the strongest connections from:", "←", incoming)}`,
      });
    }),
    usage: "/friends",
  };
}

export function statsCommands(dependencies: AppDependencies): ReadonlyArray<CommandDefinition> {
  return [
    seenCommand(dependencies),
    statsCommand(dependencies, true),
    statsCommand(dependencies, false),
    totalCommand(dependencies, true),
    totalCommand(dependencies, false),
    botStatsCommand(dependencies),
    friendsCommand(dependencies),
  ];
}
